openvr_ros_driver/src/openvr_driver.cpp

```diff
@@ -0,0 +1,360 @@
+        msg_hmd.posestamped.pose.orientation.w = quaternion.w;
+
+        // Header
+        msg_hmd.posestamped.header.stamp = ros::Time::now();
+        msg_hmd.posestamped.header.frame_id = "vive_base";
+
+        msg.hmd = msg_hmd;
+
+        break;
+      }
+
+      case vr::ETrackedDeviceClass::TrackedDeviceClass_Controller:
+      {
+        vr::VRSystem()->GetControllerStateWithPose(vr::TrackingUniverseStanding, unDevice, &controllerState, sizeof(controllerState), &trackedDevicePose);
+
+        position = getPosition(devicePose->mDeviceToAbsoluteTracking);
+        quaternion = getRotation(devicePose->mDeviceToAbsoluteTracking);
+
+        vVel = trackedDevicePose.vVelocity;
+        vAngVel = trackedDevicePose.vAngularVelocity;
+        eTrackingResult = trackedDevicePose.eTrackingResult;
+        bPoseValid = trackedDevicePose.bPoseIsValid;
+
+        vive_msgs::Controller msg_controller;
+
+        // Role
+        msg_controller.id = vr::VRSystem()->GetControllerRoleForTrackedDeviceIndex(unDevice);
+
+        // Pose
+        msg_controller.posestamped.pose.position.x = position.v[0];
+        msg_controller.posestamped.pose.position.y = position.v[1];
+        msg_controller.posestamped.pose.position.z = position.v[2];
+        msg_controller.posestamped.pose.orientation.x = quaternion.x;
+        msg_controller.posestamped.pose.orientation.y = quaternion.y;
+        msg_controller.posestamped.pose.orientation.z = quaternion.z;
+        msg_controller.posestamped.pose.orientation.w = quaternion.w;
+
+        // Buttons
+        const uint64_t buttonBitmasks[4] = {
+            (uint64_t)pow(2, (int) vr::k_EButton_ApplicationMenu),
+            (uint64_t)pow(2, (int) vr::k_EButton_Grip),
+            (uint64_t)pow(2, (int) vr::k_EButton_SteamVR_Touchpad),
+            (uint64_t)pow(2, (int) vr::k_EButton_SteamVR_Trigger)
+        };
+
+        msg_controller.joystick.buttons.resize(4);
+
+        for (int button = 0; button < 4; ++button) {
+          if ((controllerState.ulButtonPressed & buttonBitmasks[button]) != 0) {
+            msg_controller.joystick.buttons[button] = 2;
+          } else if ((controllerState.ulButtonTouched & buttonBitmasks[button]) != 0) {
+            msg_controller.joystick.buttons[button] = 1;
+          } else {
+            msg_controller.joystick.buttons[button] = 0;
+          }
+        }
+
+        // Axes
+        msg_controller.joystick.axes.push_back(controllerState.rAxis[0].x);
+        msg_controller.joystick.axes.push_back(controllerState.rAxis[0].y);
+        msg_controller.joystick.axes.push_back(controllerState.rAxis[1].x);
+
+        // Header
+        msg_controller.posestamped.header.stamp = ros::Time::now();
+        msg_controller.posestamped.header.frame_id = "vive_base";
+
+        msg.controllers.push_back(msg_controller);
+
+        break;
+      }
+
+      case vr::ETrackedDeviceClass::TrackedDeviceClass_TrackingReference:
```

Comment:
This case does not do anything, but we might want to expand this in the future. Add a comment like `//Currently tracking reference is not used`